cpp/inc/bond/comm/transport/epoxy.h

```diff
@@ -772,27 +739,29 @@ class EpoxyTransport
         EpoxyConnection(const Allocator& allocator,
                         const SocketAddress& address,
                         const boost::shared_ptr<INetworkConnectionSink>& handler,
-                        boost::asio::io_service& ioService)
+                        const boost::shared_ptr<boost::asio::io_service>& ioServicePtr)
             : m_allocator(allocator)
             , m_address(address)
+            , m_ioServicePtr(ioServicePtr)
             , m_handler(handler)
-            , m_socket(ioService)
+            , m_socket(*ioServicePtr)
```

Comment:
We're unconditionally dereferencing a `shared_ptr` that might be empty. Similar things happen throughout this PR.